src/System.Linq/tests/ToHashSetTests.cs

```diff
@@ -58,17 +59,184 @@ public void TolerateDuplicates()
+                source => source.Where(x => x < 40), // WhereArrayIterator, WhereEnumerableIterator, WhereListIterator
+                source => source.Where(x => x < 40).Select(x => x) // WhereSelectArrayIterator, WhereSelectEnumerableIterator, WhereSelectListIterator
+            );
+
+            testData.AddSources(GetSources());
+
+            return testData.GetData();
+        }
+
+        private static IEnumerable<object[]> GetGroupingData()
+        {
+            var testData = new WorkAsExpectedData<IGrouping<int, int>>();
+
+            testData.AddDefaultComparers();
+
+            testData.AddActions(
+                source => source.GroupBy(x => x), // GroupedEnumerable<TSource, TKey>, GroupedEnumerable<TSource, TKey, TElement>
+                source => source.GroupBy(x => x, x => x), // GroupedResultEnumerable<TSource, TKey, TResult>, GroupedResultEnumerable<TSource, TKey, TElement, TResult>
+                source => source.ToLookup(i => i) // Lookup
+            );
+
+            testData.AddSources(GetSources());
+
+            return testData.GetData();
+        }
+
+        /// <summary>
+        /// Assert.Equals -> HashSet.SetEquals not work for IGrouping
+        /// </summary>
+        /// <param name="expected">Expected set</param>
+        /// <param name="actual">Actual set</param>
+        private void AssertHashSetGroupingEqual(HashSet<IGrouping<int, int>> expected, HashSet<IGrouping<int, int>> actual)
+        {
+            // The Count should be equal
+            Assert.Equal(expected.Count, actual.Count);
+
+            // Get the group by key and check as IEnumerable
+            foreach (var expectedGroup in expected)
+            {
+                var group = actual.FirstOrDefault(x => x.Key == expectedGroup.Key);
+
+                Assert.NotNull(group);
+
+                Assert.Equal(expectedGroup.AsEnumerable(), group.AsEnumerable());
+            }
+        }
+        
+        private static IEnumerable<IEnumerable<int>> GetSources()
+        {
+            // Create empty, distinct and repeated sources
+            var sources = new[]
+            {
+                Array.Empty<int>(),
+                Range(0, 50),
+                Range(0, 50).Concat(Range(20, 50))
+            };
+
+            // Some methods have specific handling for Arrays, Lists and IEnumerables
+            foreach (var source in sources)
+            {
+                yield return source.ToArray();
+                yield return new List<int>(source);
+                yield return new SortedSet<int>(source);
+            }
+
+            // Return RepeatIterator and RangeIterator
+            yield return Repeat(0, 50);
+            yield return Range(0, 50);
+        }
+
+        /// <summary>
+        /// Responsible for combining Actions, Comparers and Sources
+        /// </summary>
+        /// <typeparam name="T">Result Type</typeparam>
+        private class WorkAsExpectedData<T>
```

Comment:
Nit: I found this name confusing because with xUnit `Data` usually means that the type implements `IEnumerable<object[]>` and you can return it from a MemberData method directly. Naming it something like `TestDataBuilder` or `ToHashSetDataBuilder` would be preferred.